keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,277 @@
+* creates a `deployment` of **kube-migrator controller** 
+* generates a comprehensive list of resource types via the discovery API
+* discovers all custom resources via listing CRDs
+* discovers all aggregated resources via listing all `apiservices` that have
+  `.spec.service != null`
+* removes the custom resources and aggregated resources from the comprehensive
+  resource list. The list now only contains Kubernetes built-in resources.
+* removes resources that share the same storage. At the alpha stage, the
+  information is hard-coded, like in this [list][].
+* creates `migration` CRD (see the [API section][] for the schema) if it does
+  not exist.
+* creates `migration` CRs for all remaining resources in the list. The
+  `ownerReferences` of the `migration` objects are set to the **kube-migrator
+  controller** `deployment`. Thus, the old `migration`s are deleted with the old
+  `deployment` in the first step.
+
+The control loop of **kube-migrator controller** does the following:
+* watches the instances of the `migration` CR, adds them to a [workqueue][].
+* syncs one `migration` at a time to avoid overloading the apiserver,
+  * if `migration.status` is nil, it creates a **migration worker**
+    `job` to migrate the resource type. The `ownerReference` of the `job` is set
+    to the `migration`.
+  * adds the *Running* condition to `migration.status.conditions`.
+  * waits until the **migration worker** `job` finishes, adds either the
+    *Succeeded* or *Failed* condition to `migration.status.conditions` and sets
+    the *Running* condition to false.
+
+The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
+--include=<resource type>` to migrate a resource type. The **migrator worker**
+uses API chunking to retrieve partial lists of a resource type and thus can
+migrate a small chunk at a time. It stores the [continue token] in the owner
+`migration.status.continueToken`. With the inconsistent continue token
+introduced in [#67284][], the **migrator worker** does not need to worry about
+expired continue token.
+
+[list]:https://github.com/openshift/origin/blob/2a8633598ef0dcfa4589d1e9e944447373ac00d7/pkg/oc/cli/admin/migrate/storage/storage.go#L120-L184
+[workqueue]:https://github.com/kubernetes/client-go/tree/master/util/workqueue
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#alpha-api
+
+The cluster admin can run the `kubectl wait --for=condition=Succeeded
+migrations` to wait for all migrations to succeed.
+
+Users can run `kubectl create` to create `migration`s to request migrating
+custom resources and aggregated resources.
+
+### Alpha API
+
+We introduce the `migration` API to record the intention and the progress of a
+migration. The API will be a CRD defined in the `migration.k8s.io` group.
+
+Read the [workflow section][] to understand how the API is used.
+
+```golang
+type StorageVersionMigration struct {
```

Comment:
Added a line saying that we abbreviated it as `migration` in the text.

I think the API name should be fully qualified, because we might add other kind of migrations in the future.